### ISSUE TITLE
look into / ask about megazeux update loop

### ISSUE THREAD

**Issue by increpare** [increpare/PuzzleScript#896]
I tried looking at the code myself but it was just a bit too big of a project - I should just ask :)

**Comment by zzo38** [comment]
How is the relevance of MegaZeux, in PuzzleScript?

**Comment by increpare** [comment]
Puzzlescript has functionality for realtime games - https://www.puzzlescript.net/Documentation/realtime.html , and there are several modes.  This makes it quite similar to ZZT/Megazeux.  But I've never been totally convinced by any of them as I implemented them(how the 'tick' of the natural world should combine with the input of the player), so an exact comparison with other similar engines might be helpful!

**Comment by zzo38** [comment]
I know how ZZT is working. The working of ZZT is that stats are executed in the order that they appear in the stat list, where the player is the first stat. Some stats are slow, and will not be executed on every turn; the exact criteria is `(Cycle <> 0) and ((CurrentTick mod Cycle) = (CurrentStatTicked mod Cycle))`, where `Cycle` controls the speed, and `CurrentStatTicked` is the stat nmber (zero for the player). Input is handled during the player's turn (stat 0).

MegaZeux works differently than ZZT, and does not have stats like ZZT.

**Comment by increpare** [comment]
Oh, someone who knows! Hello! 

So would you say that ZZT has a fixed timer and processes input+other events on that timer?

I'm going to past a list of timing possibilities - kind of for my own future reference - at least three of them must already be supported by puzzlescript but I'm not sure exactly which three anymore ^^

This is the legend:
![image](https://user-images.githubusercontent.com/465632/177361699-cf7b9e78-47f8-4f89-b460-149d664ca6b2.png)

In the standard turn-based mode the engine just does stuff whenever you input.
![image](https://user-images.githubusercontent.com/465632/177361750-911c7a85-a519-4491-a535-8d7918bbbaf1.png)

A naive approach to realtime timers is to just have a game timer, and have input events trigger their own events
![image](https://user-images.githubusercontent.com/465632/177362712-d6a92b6b-db34-4763-baa5-9b298eac04ca.png)

One problem with this is that maybe the player can move infinitely quickly if you press keys quickly.

One way of working with this is to have the input rate be throttled to the frame rate (even if the inputs are processed immediately, you can't trigger input events faster than the framerate)
![image](https://user-images.githubusercontent.com/465632/177363966-2b14047e-abd6-4d48-a99e-73d11593816c.png)


One can have a fixed timer, where any events you have will trigger on that timer's tick (in puzzlescript this could take the form of queuing up multiple events and processing them all, or it could take the form of just looking at the most recent input. Or just the first. I honestly don't know what'd be best, if any)
![image](https://user-images.githubusercontent.com/465632/177361885-cea1fb88-f62a-4bbf-a5a9-3b779cbdaad7.png)

An alternative to the above which might feel more responsive is the force a tick when you press a button - bringing
![image](https://user-images.githubusercontent.com/465632/177362082-a921a0ab-45a9-45ad-b660-27350ed291d3.png)

Another possibility is to poll the keys just during the regular tick - if a key not held down when the tick happens, ignore it
![image](https://user-images.githubusercontent.com/465632/177364221-03ec56bd-839d-419b-84e7-62f5380b398e.png)
(this is weird because you can quickly tap keys between frames and nothing happens)

Another options is, when you press a key, to trigger the next 'frame' immediately, but to keep the subsequent frames in sync with the original timer:
![image](https://user-images.githubusercontent.com/465632/177364483-cbf1316d-6b97-4c42-b5bd-2ff03570a19e.png)

Two things unaccounted for in the above list:
1 - how they handle 'again' rules.  It's hairy and makes things much more complicated.  Are again animations blocking of the regular timer? Are they blocking of future input if it's a realtime game? Who knows!
2 - how should key-repeat work?  Well you can use the system key-repeat timer. But maybe it's better to have it synced to the engine ticks right?
3 - how you detect if you're on an input event frame vs a normie frame, and what games benefit from what systems.

**Comment by zzo38** [comment]
> So would you say that ZZT has a fixed timer and processes input+other events on that timer?

Yes, that is how it is working. (You can adjust the game speed by pushing S on the title screen; some forks also allow typing `?SPEED` during the game to adjust the speed.)

**Comment by increpare** [comment]
Ah thank you! Good to know!
